File: backend/app/models/order.rb
class Order < ApplicationRecord
  has_many :order_items, dependent: :destroy
  has_many :productos, through: :order_items
  # Scope para órdenes pendientes huérfanas (sin usuario) mayores a 30 minutos
  scope :orphaned_pending_orders, -> {
    where(status: "pending", user_id: nil)
      .where("created_at < ?", 30.minutes.ago)
  }

  # Método para limpiar órdenes huérfanas
  def self.cleanup_orphaned_orders
    orphaned_pending_orders.destroy_all
  end
  validates :customer_address, presence: true, if: -> { delivery_type == "domicilio" }
  validates :customer_city, presence: true, if: -> { delivery_type == "domicilio" }
  belongs_to :user, optional: true

  # Validaciones
  validates :customer_name, presence: true
  validates :customer_email, presence: true, format: { with: URI::MailTo::EMAIL_REGEXP }
  validates :customer_phone, presence: true
  validates :total_amount, presence: true, numericality: { greater_than: 0 }

  # Enums para el estado del pedido
  enum :status, {
    pending: "pending",      # Pendiente
    processing: "processing",     # En proceso
    paid: "paid",          # Finalizado
    failed: "failed",        # Pago fallido
    cancelled: "cancelled"       # Cancelado
  }

  # Enum para tipo de entrega
  enum :delivery_type, {
    domicilio: "domicilio",
    recoger: "recoger"
  }, prefix: true

  # Devuelve un mensaje según el estado y tipo de entrega
  def estado_entrega_mensaje
    if status == "paid" || status == "processing"
      if delivery_type_domicilio?
        "Su pedido va en camino. Tiempo estimado: #{estimated_time || 30} minutos."
      elsif delivery_type_recoger?
        "Puede venir por su pedido."
      end
    else
      "Su pedido está en proceso."
    end
  end

  # Enums para método de pago
  enum :payment_method, {
    credit_card: "credit_card",
    debit_card: "debit_card",
    pse: "pse",                  # PSE (Pagos Seguros en Línea)
    cash: "cash"
  }

  # Callbacks
  before_create :generate_order_number
  after_create :create_order_items_from_cart

  # ✅ Nuevo callback: enviar email cuando el pedido se marque como finalizado
  after_update :send_finalized_email, if: -> { saved_change_to_status? && status.to_s.strip.downcase == 'paid' }

  # Métodos de instancia
  def add_product(producto, quantity = 1)
    order_items.create!(
      producto: producto,
      quantity: quantity,
      unit_price: producto.precio || 0,
      total_price: (producto.precio || 0) * quantity
    )
  end

  def update_total
    update(total_amount: order_items.sum(:total_price))
  end

  def can_be_paid?
    status_pending? && total_amount > 0
  end

  def mark_as_paid!(payu_transaction_id, payu_response)
    update!(
      status: :processing, # Ahora queda en proceso después del pago
      payu_transaction_id: payu_transaction_id,
      payu_response: payu_response
    )
  end

  def mark_as_failed!(payu_response)
    update!(
      status: :failed,
      payu_response: payu_response
    )
  end

  private

  def generate_order_number
    loop do
      self.order_number = "ORD-#{Time.current.strftime('%Y%m%d')}-#{SecureRandom.hex(4).upcase}"
      break unless Order.exists?(order_number: order_number)
    end
  end

  def create_order_items_from_cart
    # Este método se implementará cuando integremos con el carrito
    # Por ahora está vacío
  end

  # ✅ Método privado para enviar el email de factura cuando el pedido finaliza
  def send_finalized_email
    CustomerInvoiceMailer.invoice(self).deliver_later
  rescue => e
    Rails.logger.error("[Order#send_finalized_email] order_id=#{id} error=#{e.class}: #{e.message}")
  end
end

File: frontend/zona44app/lib/models/order.dart
class Order {
  final int id;
  final String orderNumber;
  final String status;
  final String deliveryType;
  final double totalAmount;
  final DateTime? createdAt;
  final List<OrderItem> orderItems;
  final OrderUser? user;
  // Nuevos campos para datos de cliente sin usuario autenticado
  final String customerName;
  final String customerEmail;
  final String customerPhone;
  final String customerAddress;
  final String customerCity;

  Order({
    required this.id,
    required this.orderNumber,
    required this.status,
    required this.deliveryType,
    required this.totalAmount,
    required this.createdAt,
    required this.orderItems,
    this.user,
    this.customerName = '',
    this.customerEmail = '',
    this.customerPhone = '',
    this.customerAddress = '',
    this.customerCity = '',
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    DateTime? createdAt;
    try {
      if (json['created_at'] != null &&
          json['created_at'].toString().isNotEmpty) {
        createdAt = DateTime.tryParse(json['created_at'].toString());
      }
    } catch (_) {
      createdAt = null;
    }
    return Order(
      id: json['id'] ?? 0,
      orderNumber: json['order_number']?.toString() ?? '',
      status: json['status']?.toString() ?? 'pending',
      deliveryType: json['delivery_type']?.toString() ?? '',
      totalAmount: (json['total_amount'] is num)
          ? (json['total_amount'] as num).toDouble()
          : double.tryParse(json['total_amount']?.toString() ?? '') ?? 0.0,
      createdAt: createdAt,
      orderItems:
          (json['order_items'] as List?)
              ?.map((item) => OrderItem.fromJson(item))
              .toList() ??
          [],
      user: json['user'] != null ? OrderUser.fromJson(json['user']) : null,
      customerName: json['customer_name']?.toString() ?? '',
      customerEmail: json['customer_email']?.toString() ?? '',
      customerPhone: json['customer_phone']?.toString() ?? '',
      customerAddress: json['customer_address']?.toString() ?? '',
      customerCity: json['customer_city']?.toString() ?? '',
    );
  }
}

class OrderUser {
  final int id;
  final String email;
  final String firstName;
  final String lastName;

  OrderUser({
    required this.id,
    required this.email,
    required this.firstName,
    required this.lastName,
  });

  String get fullName => '$firstName $lastName';

  factory OrderUser.fromJson(Map<String, dynamic> json) {
    return OrderUser(
      id: json['id'] ?? 0,
      email: json['email']?.toString() ?? '',
      firstName: json['first_name']?.toString() ?? '',
      lastName: json['last_name']?.toString() ?? '',
    );
  }
}

class OrderItem {
  final int id;
  final int quantity;
  final double unitPrice;
  final double totalPrice;
  final Producto producto;

  OrderItem({
    required this.id,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
    required this.producto,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    // Debug: Imprimir el JSON del item
    print('🔍 OrderItem JSON: $json');

    return OrderItem(
      id: json['id'] ?? 0,
      quantity: json['quantity'] ?? 0,
      unitPrice: (json['unit_price'] is num)
          ? (json['unit_price'] as num).toDouble()
          : double.tryParse(json['unit_price']?.toString() ?? '') ?? 0.0,
      totalPrice: (json['total_price'] is num)
          ? (json['total_price'] as num).toDouble()
          : double.tryParse(json['total_price']?.toString() ?? '') ?? 0.0,
      producto: json['producto'] != null
          ? Producto.fromJson(json['producto'])
          : Producto.empty(),
    );
  }
}

class Producto {
  final int id;
  final String name;
  final double precio;
  final String descripcion;

  Producto({
    required this.id,
    required this.name,
    required this.precio,
    required this.descripcion,
  });

  factory Producto.fromJson(Map<String, dynamic> json) {
    // Debug: Imprimir el JSON del producto
    print('🔍 Producto JSON: $json');

    return Producto(
      id: json['id'] ?? 0,
      name: json['name']?.toString() ?? '',
      precio: (json['precio'] is num)
          ? (json['precio'] as num).toDouble()
          : double.tryParse(json['precio']?.toString() ?? '') ?? 0.0,
      descripcion: json['descripcion']?.toString() ?? '',
    );
  }

  factory Producto.empty() =>
      Producto(id: 0, name: '', precio: 0.0, descripcion: '');
}
